packages/protocol/contracts/stability/StableToken.sol

```diff
@@ -519,4 +522,48 @@ contract StableToken is
     return true;
   }
 
+  /**
+   * @notice Reserve balance for making payments for gas in this StableToken currency.
+   * @param from The account to reserve balance from
+   * @param value The amount of balance to reserve
+   */
+  function reserveGas(address from, uint256 value)
+    external
+    onlyVm
+    onlyWhenNotFrozen
+    updateInflationFactor
+  {
+    uint256 units = _valueToUnits(inflationState.factor, value);
+    totalSupply_ = totalSupply_.sub(units);
+    balances[from] = balances[from].sub(units);
+  }
+
+  /**
+   * @notice Refund balance after making payments for gas in this StableToken currency.
+   * @param to The account to credit balance to
+   * @param value The amount of balance to refund
+   * @dev We can assume that the inflation factor is up to date as `reserveGas`
+   * will have been called in the same transaction
+   */
+  function refundGas(address to, uint256 value) external onlyVm onlyWhenNotFrozen {
```

Comment:
One issue with refunding and crediting in separate functions is that there's no guarantee that the transaction does not change the total supply due to rounding errors.

I don't think this is critical, but would be curious to get your thoughts on the tradeoffs between this and the approach I suggested

Reply:
About rounding error, the current version is the same as the previous one (actually there isn't much difference now).
Your approach would move some logic from celo-blockchain to the contracts, I think that is good, but would prefer that same would then work for gold token too.

Reply:
Good point. I'm happy with whatever you decide is best